### PR TITLE
Add SseFormatter

### LINKED ISSUE #109294: [API Proposal]: Add formatting support to `System.Net.ServerSentEvents`
### Background and motivation

The `System.Net.ServerSentEvents` library added support for parsing SSE events. We should also add support for formatting SSE events on the server side, in a way that unlocks work for https://github.com/dotnet/aspnetcore/issues/56172.

### API Proposal

```diff
namespace System.Net.ServerSentEvents;

// Extends the existing type to support writing
public readonly struct SseItem<T>
{
-   public SseItem(T data, string eventType);
+   public SseItem(T data, string? eventType);

    public T Data { get; }

-   public string EventType { get; }
+   public string? EventType { get; } 

+   public string? EventId { get; init; }
}

public delegate T SseItemParser<out T>(string eventType, ReadOnlySpan<byte> data);
+public delegate byte[] SseItemFormatter<T>(T value);

+public sealed class SseFormatter
+{
+    public static SseFormatter Create<T>(IAsyncEnumerable<SseItem<string>> source);
+    public static SseFormatter Create<T>(IAsyncEnumerable<SseItem<T>> source, SseItemFormatter<T> itemFormatter);
+
+    public Task WriteToStreamAsync(Stream stream, CancellationToken cancellationToken = default);
+    public void WriteToStream(Stream stream, CancellationToken cancellationToken = default);
+}
```

### API Usage

```csharp
IAsyncEnumerable<SseItem<int>> source = ...;
SseFormatter formatter = SseFormatter.Create<int>(source, value => JsonSerializer.SerializeToBytes(value));
await formatter.WriteToStreamAsync(targetStream, cancellationToken);
```

### Alternative Designs

The current SseItemFormatter delegate uses an allocating design, but which is simpler to use. Alternative designs would involve writing either to a `Stream`:

```csharp
public delegate ValueTask SseItemFormatter<T>(Stream stream, T value);
```

But this would require wrapping the underlying stream in order to intercept potential line breaks. Alternatively, we could use an `IBufferWriter<byte>` which also composes better with `PipeWriter` targets:

```csharp
public delegate void SseItemFormatter<T>(IBufferWriter<byte> writer, T value);
```

The downside of either approach is that it becomes harder to define formatters from the user's perspective:

```csharp
SseFormatter formatter = SseFormatter.Create<int>(source, (bufferWriter, value) => 
{ 
    var writer = new Utf8JsonWriter(bufferWriter);
    return JsonSerializer.SerializeToBytes(bufferWriter, value); 
});
```

### Risks

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #109294.

### CODE REVIEW

File: src/libraries/System.Net.ServerSentEvents/src/Resources/Strings.resx [hunk 1]
```diff
@@ -120,4 +120,7 @@
   <data name="InvalidOperation_EnumerateOnlyOnce" xml:space="preserve">
     <value>The enumerable may be enumerated only once.</value>
   </data>
+  <data name="ArgumentException_MustNotContainLineBreaks" xml:space="preserve">
+    <value>Parameter should not contain any line breaks.</value>
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/Resources/Strings.resx:124]:
```suggestion
    <value>Argument must not contain line breaks.</value>
```

File: src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs [hunk 2]
```diff
@@ -6,14 +6,20 @@
 
 namespace System.Net.ServerSentEvents
 {
+    public static partial class SseFormatter
+    {
+        public static System.Threading.Tasks.Task WriteAsync(System.Collections.Generic.IAsyncEnumerable<System.Net.ServerSentEvents.SseItem<string>> source, System.IO.Stream destination, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
+        public static System.Threading.Tasks.Task WriteAsync<T>(System.Collections.Generic.IAsyncEnumerable<System.Net.ServerSentEvents.SseItem<T>> source, System.IO.Stream destination, System.Action<System.Buffers.IBufferWriter<byte>, System.Net.ServerSentEvents.SseItem<T>> itemFormatter, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: This ordering of IBufferWriter then SseItem looks wrong to me. We generally do source and then destination; the source here is the SseItem and the destination is the IBufferWriter.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseItem.cs [hunk 3]
```diff
@@ -9,12 +9,16 @@ public readonly struct SseItem<T>
...
+        /// <summary>The event's id.</summary>
+        private readonly string? _eventId;
 
         /// <summary>Initializes the server-sent event.</summary>
         /// <param name="data">The event's payload.</param>
         /// <param name="eventType">The event's type.</param>
-        public SseItem(T data, string? eventType)
+        public SseItem(T data, string? eventType = null)
         {
+            Helpers.ValidateParameterDoesNotContainLineBreaks(eventType, nameof(eventType));
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseItem.cs:20]: Can you add an XML comment for the exception?

File: src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: I was thinking of how serializers typically order their paramers, e.g. JsonConverter.Write accepts writer first and value second.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseItem.cs (hunk 3 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseItem.cs:21]: Done.

File: src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: cc: @bartonjs, @terrajobst

Comment (bartonjs) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: I believe that we intentionally flipped the order in the callback to be (source, destiantion).

Notably, your PR here does not match what was approved.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: > Notably, your PR here does not match what was approved.

Yep, this is exploring possible amendments in the callback. Another one is passing the wrapping SseItem<T> instead of just T. Will notify fxdc once we're settled on the shape.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: > I believe that we intentionally flipped the order in the callback to be (source, destiantion).

While technically the value is the source and the writer is the destination, this ordering is not what we commonly use in the context of serialization. For example:

* [JsonConverter.Write(Utf8JsonWriter, T)](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter-1.write?view=net-8.0)
* [JsonSerializer.SerializeAsync(Stream, T)](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.serialize?view=net-8.0#system-text-json-jsonserializer-serialize(system-io-stream-system-object-system-type-system-text-json-serialization-jsonserializercontext))
* [DataContractSerializer.WriteObject(XmlWriter, object)](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.serialization.datacontractserializer.writeobject?view=netframework-4.8.1#system-runtime-serialization-datacontractserializer-writeobject(system-xml-xmlwriter-system-object))

Given that the callback does represent a serialization callback, I would suggest we use the order as presented in this PR.

Comment (bartonjs) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs:12]: https://youtu.be/ehUiLj2KvCg?t=5583

It started as writer first, then got flipped.  There was no objection.  And, wow, I hate listening to myself.

If you want to change it, you can bring it back for discussion.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs]: https://github.com/dotnet/runtime/pull/109832/commits/1b7a3e534a9d8a2f15ac155c2352c37740da3346

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+                static async ValueTask WriteAsyncUsingPooledBuffer(Stream stream, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    byte[] sharedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length);
+                    buffer.Span.CopyTo(sharedBuffer);
+                    try
+                    {
+                        await stream.WriteAsync(sharedBuffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false);
+                    }
+                    finally
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: > The array pool should handle dropped arrays in case of an exception.

How is that?

[the remark this answers is not in the document]
Comment (gfoidl) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: In case of an exception the array won't be returned to the pool, so ultimately it will be GCed.
That's no problem for the array pool, so the `try-finally` isn't needed.

File: src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs]: I don't feel strongly about it; there are arguments to be made either way. The most important thing would be local consistency, meaning what do we expect a developer to do with the callback. If the 99% case is they're going to use JsonSerializer.Serialize in the callback, then having the order of arguments to the callback match that of the order of the method the callback will invoke seems reasonable. It'd be a stronger argument if you could just use JsonSerializer.Serialize as the callback (just passing the method group), but that won't work because it doesn't have an `IBufferWriter<byte>` overload and (with your changes) this is passing an `SseItem<T>` rather than a `T`.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/ref/System.Net.ServerSentEvents.cs]: I think there's an argument to be made in principle, that the `IBufferWriter<byte>` parameter represents a stateful writer as opposed to a destination buffer. We typically place stateful writers as the first parameter and the value to be written as the second parameter. I won't die on this hill though.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gfoidl) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: The exception here isn't the common case, so one can avoid additional complexity brought by try-finally that is not needed.

The pattern with try-finally for the array pool was used at the beginning of AP, but recently renting / returning is just done w/o try-finally. See e.g. https://github.com/dotnet/runtime/issues/71249#issuecomment-1165486832 and lots of other comments for this.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: I don't think inserting a try/finally is so terrible from a code complexity standpoint. We use that pattern pervasively and I don't think we need to remove it here.

[the remark this answers is not in the document]
Comment (gfoidl) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: > We use that pattern pervasively

Really? Recently I see it the other way around.

But to shorten this discussion to whatever you think is right here.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: I don't at all agree with that approach. Should we be starving the buffer pool just because of a transient network outage?

[the remark this answers is not in the document]
Comment (bartonjs) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: async operations aren't necessarily done referencing the buffer when they throw (it's possible the native half can write after the managed half throws a timeout exception, for example), so they're not supposed to return it except in successful flows.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: > it's possible the native half can write after the managed half throws a timeout exception, for example

Can you give an example in the repo where that might happen?

[the remark this answers is not in the document]
Comment (bartonjs) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/Helpers.cs:82]: Not off the top of my head.  I remember it being A Thing that Levi was pushing in 2919 or earlier (I see a commit I made in 2019 getting rid of a bunch of async finally-returns); but I can't find the written version of that.

Something involving patterns where you call a native equivalent of BeginRead(state, dest, onEndCallback), then Task cancellation tries to cancel the OS async operation, but that causes a race condition with the IO completion event and the async operation cancellation event.

@GrabYourPitchforks might be able to spout one off... or perhaps correct my mis-saved "fact"

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs [hunk 5]
```diff
@@ -119,6 +122,26 @@ private static void FormatSseEvent(
...
+                long retryMs = (long)retry.TotalMilliseconds;
+
+                bufferWriter.WriteUtf8String("retry: "u8);
+#if NET
+                const int MaxDigits = 20;
+                Span<byte> buffer = stackalloc byte[MaxDigits];
+                bool success = retryMs.TryFormat(buffer, out int bytesWritten, provider: CultureInfo.InvariantCulture);
+                Debug.Assert(success);
+
+                bufferWriter.Write(buffer.Slice(0, bytesWritten));
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs:138]: Is writing into a stackalloc'd span better than:
```C#
const int MaxDigits = 20;
Span<byte> buffer = bufferWriter.GetSpan(MaxDigits);
bool success = retryMs.TryFormat(buffer, out int bytesWritten, provider: CultureInfo.InvariantCulture);
Debug.Assert(success);
bufferWriter.Advance(bytesWritten);
```
?

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs:138]: Good idea, I'll use that instead.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs [hunk 6]
```diff
@@ -418,7 +425,7 @@ private bool ProcessLine(out SseItem<T> sseItem, out int advance)
 #endif
                     NumberStyles.None, CultureInfo.InvariantCulture, out long milliseconds))
                 {
-                    ReconnectionInterval = TimeSpan.FromMilliseconds(milliseconds);
+                    _nextReconnectionInterval = ReconnectionInterval = TimeSpan.FromMilliseconds(milliseconds);
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:428]: Do we need to be concerned about the exception that'll happen if this overflows?

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:428]: Good question. Not sure what the best approach is here. We could either:

1. Throw a better exception message,
2. Normalize to `TimeSpan.MaxValue` or
3. Ignore the field altogether

Seems like option 3 is more compatible with SSE's general approach to handling invalid data.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs (hunk 5 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs]: Done.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs (hunk 6 above)
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs:428]: Sounds good.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseParser_1.cs]: Done.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+            _rentedBuffer.AsSpan(0, _index).Clear();
+            _index = 0;
+        }
+
+        // Returns the rented buffer back to the pool
+        public void Dispose()
+        {
+            if (_rentedBuffer == null)
+            {
+                return;
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:68]: Do we expect this to ever happen? Can we Debug.Fail if it does?

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                return _rentedBuffer.Length;
+            }
+        }
+
+        public void Reset()
+        {
+            Debug.Assert(_rentedBuffer != null);
+            Debug.Assert(_index <= _rentedBuffer.Length);
+
+            _rentedBuffer.AsSpan(0, _index).Clear();
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:59]: Why is this needed?

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+
+namespace System.Net.ServerSentEvents
+{
+    internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
```
Comment (MihaZupan) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: Could this use https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Net/ArrayBuffer.cs?
(looks like the same idea at a quick glance)

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        {
+            // Writes a potentially multi-line string, prefixing each line with the given prefix.
+            // Both \n and \r\n sequences are normalized to \n.
+
+            while (true)
+            {
+                writer.WriteUtf8String(prefix);
+
+                int i = data.IndexOf((byte)'\n');
+                if (i is -1)
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs:146]:
```suggestion
                if (i < 0)
```

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        private static void WriteLinesWithPrefix(PooledByteBufferWriter writer, ReadOnlySpan<byte> prefix, ReadOnlySpan<byte> data)
+        {
+            // Writes a potentially multi-line string, prefixing each line with the given prefix.
+            // Both \n and \r\n sequences are normalized to \n.
+
+            while (true)
+            {
+                writer.WriteUtf8String(prefix);
+
+                int i = data.IndexOf((byte)'\n');
```
Comment (stephentoub) [inline src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs:145]: The spec implies that \r, \n, or \r\n should be used as line separators, but this is only looking for \n and \r\n.
```
Lines must be separated by either a U+000D CARRIAGE RETURN U+000A LINE FEED (CRLF) character pair, a single U+000A LINE FEED (LF) character, or a single U+000D CARRIAGE RETURN (CR) character.
```

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 9 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: Looks similar, but AFAICT it doesn't implement IBufferWriter. FWIW this is duplicated from STJ so clearly we would benefit from a shared thing.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:59]: It's following practice used in STJ and specifically the PooledByteBufferWriter that it got copied from.

Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:59]: Is it clearing buffers before returning them not considered good practice?

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 7 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:68]: We can. This is just copying from the class in STJ.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+                int newSize = currentLength + growBy;
+
+                if ((uint)newSize > MaximumBufferSize)
+                {
+                    newSize = currentLength + sizeHint;
+                    if ((uint)newSize > MaximumBufferSize)
+                    {
+                        Throw();
+                        static void Throw() => throw new OutOfMemoryException();
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:129]: The STJ variant uses a ThrowHelper method, I figured I could just replace it with a local. Are you suggesting that it should be moved to `ThrowHelper` or that factoring the throw into a method isn't necesasry?

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 8 above)
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:59]: > Is it clearing buffers before returning them not considered good practice?

a) This isn't returning it. It's resetting it for the next write.
b) We generally don't on return except for specific circumstances, like being in a library like crypto that wants extra scrutiny or returning arrays of reference types to avoid rooting objects that may be stored in the array. The default for Return is to not clear. For example, the WriteAsync polyfill earlier is not clearing.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 9 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: Yep. The STJ type implements `PipeWriter` instead of `IBufferWriter` so we can't just share the source without taking an extra dependency in pre 9.0 targets.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:129]: It's not necessary and just makes the code more convoluted.

We also should avoid copying such a large body of code. Let's separate out the STJ utility class into something that can be built into this library as well.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 9 above)
Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: > Yep. The STJ type implements `PipeWriter` instead of `IBufferWriter` so we can't just share the source without taking an extra dependency in pre 9.0 targets.

And that's not a case of just having two partial files, one with this type implementing IBufferWriter and one with it also adding the PipeWriter derivation?

Comment (MihaZupan) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: What I had in mind re: `ArrayBuffer` is if we can replace this file with something along the lines of
```c#

internal sealed class PooledByteBufferWriter : IBufferWriter<byte>, IDisposable
{
	private ArrayBuffer _buffer = new(initialSize: 256, usePool: true);
	
	public void Advance(int count) => _buffer.Commit(count);

	public Memory<byte> GetMemory(int sizeHint = 0)
	{
		_buffer.EnsureAvailableSpace(sizeHint);
		return _buffer.AvailableMemory;
	}

	public Span<byte> GetSpan(int sizeHint = 0)
	{
		_buffer.EnsureAvailableSpace(sizeHint);
		return _buffer.AvailableSpan;
	}
	
	public ReadOnlyMemory<byte> WrittenMemory => _buffer.ActiveMemory;
	
	public int Capacity => _buffer.Capacity;
	
	public int WrittenCount => _buffer.ActiveLength;
	
	public void Reset() => _buffer.Discard(_buffer.ActiveLength);
	
	public void Dispose() => _buffer.Dispose();
}
```

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 7 above)
Comment (bartonjs) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:68]: In general Dispose is supposed to be safe to call multiple times.  Are you just asserting it isn't going to happen in implementation with this internal type?

Comment (stephentoub) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:68]: > Are you just asserting it isn't going to happen in implementation with this internal type?

Yes. It's internal and not being exposed externally. It'd be misuse to dispose it multiple times.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 9 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs:9]: Thanks, this is a huge improvement.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs (hunk 11 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs]: Fixed.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs]: Removed.

File: src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs (hunk 8 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/PooledByteBufferWriter.cs]: I've refactored the code following Miha's suggestion, but it seems STJ is susceptible to the same issue. Effectively, the buffer is being cleared every time we flush to the underlying stream.

https://github.com/dotnet/runtime/blob/d4516bf13c0a06d955912c15e71a13e3dc894db0/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs#L108

I'll file an issue.

**Review by stephentoub — APPROVED** [review]
